src/transforms/aggregate.rs

```diff
@@ -0,0 +1,397 @@
+use crate::{
+    internal_events::{AggregateEventRecorded, AggregateFlushed, AggregateUpdateFailed},
+    transforms::{
+        TaskTransform,
+        Transform,
+    },
+    config::{DataType, GlobalOptions, TransformConfig, TransformDescription},
+    event::{
+        metric,
+        Event,
+        EventMetadata,
+    },
+};
+use async_stream::stream;
+use futures::{stream, Stream, StreamExt};
+use serde::{Deserialize, Serialize};
+use std::{
+    collections::HashMap,
+    pin::Pin,
+    time::{Duration},
+};
+
+#[derive(Deserialize, Serialize, Debug, Default, Clone)]
+#[serde(deny_unknown_fields, default)]
+pub struct AggregateConfig {
+    /// The interval between flushes in milliseconds.
+    #[serde(default = "default_interval_ms")]
+    pub interval_ms: u64,
+}
+
+fn default_interval_ms() -> u64 {
+    10 * 1000
+}
+
+inventory::submit! {
+    TransformDescription::new::<AggregateConfig>("aggregate")
+}
+
+impl_generate_config_from_default!(AggregateConfig);
+
+#[async_trait::async_trait]
+#[typetag::serde(name = "aggregate")]
+impl TransformConfig for AggregateConfig {
+    async fn build(&self, _globals: &GlobalOptions) -> crate::Result<Transform> {
+        Aggregate::new(self).map(Transform::task)
+    }
+
+    fn input_type(&self) -> DataType {
+        DataType::Metric
+    }
+
+    fn output_type(&self) -> DataType {
+        DataType::Metric
+    }
+
+    fn transform_type(&self) -> &'static str {
+        "aggregate"
+    }
+}
+
+type MetricEntry = (metric::MetricData, EventMetadata);
+
+//------------------------------------------------------------------------------
+
+#[derive(Debug)]
+pub struct Aggregate {
+    interval: Duration,
+    map: HashMap<metric::MetricSeries, MetricEntry>,
+}
+
+impl Aggregate {
+    pub fn new(config: &AggregateConfig) -> crate::Result<Self> {
+        Ok(Self {
+            interval: Duration::from_millis(config.interval_ms),
+            map: HashMap::new(),
+        })
+    }
+
+    fn record(&mut self, event: Event) {
+        let (series, data, metadata) = event.into_metric().into_parts();
+
+        match data.kind {
+            metric::MetricKind::Incremental => {
+                self.map.entry(series)
+                    .and_modify(|existing| {
+                        if ! existing.0.update(&data) {
+                            emit!(AggregateUpdateFailed);
+                        }
+                        existing.1.merge(metadata.clone());
+                    }).or_insert((data, metadata));
```

Comment:
The other pattern here, to avoid the `clone`, would be a match:
```suggestion
                match self.map.entry(series) {
                    hash_map::Occupied(existing) => {
                        if ! existing.0.update(&data) {
                            emit!(AggregateUpdateFailed);
                        }
                        existing.1.merge(metadata);
                    }
                    hash_map::Vacant(entry) => entry.insert((data, metadata)),
                }
```